students/jeff_shabani/session07/html_render.py

```diff
@@ -3,26 +3,171 @@
 """
 A class-based system for rendering html.
 """
-
+import functools
 
 # This is the framework for the base class
 class Element(object):
+    indent = ' ' * 2
+    tag = 'html'
+
+    def __init__(self, content=None, **attrs):
+        self.attrs = attrs
+        if content:
+            self.content = [content]
+        else:
+            self.content = []
+
+    def _front_tag(self):
+        """Creates the front/opening tag"""
+        return f'<{self.tag}>'
+
+    def _end_tag(self):
+        """Creates the ending tag"""
+        return f'</{self.tag}>'
+
+    def append(self, new_content):
+        self.content.append(new_content)
+
+    def render(self, file_name, cur_indent=''):
+        # Writes the opening tag
+        file_name.write(f'{self._front_tag()}\n')
+
+        for content_line in self.content:
+            if hasattr(content_line, 'render'):
+                content_line.render(file_name)
+            else:
+                file_name.write(f'{content_line}\n')
+        file_name.write(f'{self._end_tag()}\n')
+
+
+class Html(Element):
+    """
+    html sublcass
+    """
+    tag = 'html'
+
+    def render(self, file_name):
+        file_name.write(f'<!DOCTYPE html>\n')
+        super().render(file_name)
+
+
+class P(Element):
+    """
+    class for p tag
+    """
+    tag = 'p'
+
+
+class Body(Element):
+    """
+    class for p tag
+    """
+    tag = 'body'
+
+
+"""
+Step 3: print on one line
+"""
+class OneLineTag(Element):
+
+    tag = 'Title'
 
-    def __init__(self, content=None):
-        self.content = str(content)
+    def render(self, file_name, cur_indent=''):
+        """
+        Renders elements on a single line.
+        :param file_name:
+        :param cur_indent:
+        """
+        file_name.write(f'{self._front_tag()} ')
+        for k, v in self.attrs.items():
+            file_name.write(f'{k}="{v}"')
+        file_name.write(f'{self._front_tag()}')
```

Comment:
This one will cause the front_tag to be duplicated. Instead, add the closing ">" of the front_tag here:

```
file_name.write(f'>')
```